monai/inferers/inferer.py

```diff
@@ -64,6 +67,190 @@ def __call__(self, inputs: torch.Tensor, network: Callable, *args: Any, **kwargs
+            ratio = self._get_ratio(in_patch, out_patch)
+            ratios.append(ratio)
+            if self.splitter is None:
+                merger.initialize()
+            else:
+                output_shape = self._get_output_shape(inputs, out_patch, ratio)
+                merger.initialize(output_shape)
+        return ratios
+
+    def _aggregate(self, outputs, locations, ratios, batch_size):
+        for merger, output_patches, ratio in zip(self.mergers, outputs, ratios):
+            # split batched output into individual patches and then aggregate
+            for in_loc, out_patch in zip(locations, torch.chunk(output_patches, batch_size)):
+                out_loc = [round(l * r) for l, r in zip(in_loc, ratio)]
+                merger.aggregate(out_patch, out_loc)
+
+    def _get_ratio(self, in_patch, out_patch):
+        """Define the shape of output merged tensors"""
+        return tuple(op / ip for ip, op in zip(in_patch.shape[2:], out_patch.shape[2:]))
+
+    def _get_output_shape(self, inputs, out_patch, ratio):
+        """Define the shape of output merged tensors"""
+        in_spatial_shape = inputs.shape[2:]
+        out_spatial_shape = tuple(round(s * r) for s, r in zip(in_spatial_shape, ratio))
+        output_shape = out_patch.shape[:2] + out_spatial_shape
+        return output_shape
+
+    def __call__(
+        self,
+        inputs: torch.Tensor,
+        network: Callable[..., torch.Tensor | Sequence[torch.Tensor] | dict[Any, torch.Tensor]],
+        *args: Any,
+        **kwargs: Any,
+    ) -> torch.Tensor | Sequence[torch.Tensor] | dict[Any, torch.Tensor]:
+        """
+        Args:
+            inputs: input data for inference, a torch.Tensor, representing an image or batch of images.
+                However if the data is already split, it can be fed by providing a list of tuple (patch, location),
+                or a MetaTensor that has metadata for `PatchKeys.LOCATION`. In both cases no splitter should be provided.
+            network: target model to execute inference.
+                supports callables such as ``lambda x: my_torch_model(x, additional_config)``
+            args: optional args to be passed to ``network``.
+            kwargs: optional keyword args to be passed to ``network``.
+
+        """
+        is_merger_initialized = False
+        ratios = []
+        patch_locations: Iterable[tuple[torch.Tensor, Sequence[int]]]
+        if self.splitter is None:
+            patch_locations = inputs
+        else:
+            patch_locations = self.splitter(inputs)
+        for patches, locations, batch_size in self._batch_sampler(patch_locations):
+            # run inference
+            outputs = self._run_inference(network, patches, *args, **kwargs)
+            # initialize the mergers
+            if not is_merger_initialized:
+                ratios = self._initialize_mergers(inputs, outputs, patches, batch_size)
+                is_merger_initialized = True
```

Comment:
May I know why we don't always call `_initialize_mergers` here?
This `is_merger_initialized` is not thread-safe.

Thanks.

Reply:
`_initialize_mergers` only need to be called for the first iteration of this loop. Since threading is happing in `dataloader` (or anywhere before calling `PatchInferer`) The entire loop will be run on a single thread, so `is_merger_initialized` should not make it thread unsafe.
Having said that you are right that it is NOT thread safe, and that is because I am using `Merger` object as the input instead of `Merger` class to be instantiated by each thread (but with classes passing arguments is cumbersome). I'll see how to address this but that was a good point anyways. Thanks

Reply:
No `PatchInferer` accepts `merger_cls` (Merger class) instead of `merger` (Merger object), so each call to `PatchInferer` will instantiate a new merger. This should make `PatchInferer` thread safe.